### ISSUE TITLE
MWOffliner fails with a "RangeError: Maximum call stack size exceeded"

### ISSUE THREAD

**Issue by mgautierfr** [openzim/mwoffliner#54]
Trying to grab fr.wikiquote.org, mwoffliner fails with :

```
Getting article from https://fr.wikiquote.org/w/api.php?action=visualeditor&format=json&paction=parse&page=S%C3%A9rie_B&oldid=243531
RangeError: Maximum call stack size exceeded
    at RegExp.[Symbol.replace] (native)
    at RegExp.[Symbol.replace] (native)
    at String.replace (native)
    at Object.exports.toASCIILowerCase (/home/mgautier/Project/KIWIX/node_modules/domino/lib/utils.js:72:12)
    at HTMLAnchorElement.getAttribute (/home/mgautier/Project/KIWIX/node_modules/domino/lib/Element.js:371:36)
    at rewriteUrl (/home/mgautier/Project/KIWIX/node_modules/mwoffliner/lib/mwoffliner.lib.js:1061:40)
    at /home/mgautier/Project/KIWIX/node_modules/async/lib/async.js:181:20
    at replenish (/home/mgautier/Project/KIWIX/node_modules/async/lib/async.js:319:21)
    at /home/mgautier/Project/KIWIX/node_modules/async/lib/async.js:326:29
    at /home/mgautier/Project/KIWIX/node_modules/async/lib/async.js:44:16
```

The command I run is :
`node node_modules/mwoffliner/bin/mwoffliner.script.js --mwUrl https://fr.wikiquote.org/ --adminEmail user@exemple.fr --outputDirectory ~/Project/KIWIX/wikiquote.fr --redisSocket /tmp/redis.sock --keepHtml --verbose --cacheDirectory ~/Project/KIWIX/wikiquote.fr.cache --tmpDirectory ~/Project/KIWIX/wikiquote.fr.tmp`

mwoffliner was installed with npm and is in version 1.1.3.

**Comment by mgautierfr** [comment]
The full trace of the execution.
[mwoffliner_fr.wikiquote.org_trace.txt](https://github.com/kiwix/mwoffliner/files/872549/mwoffliner_fr.wikiquote.org_trace.txt)

**Comment by kelson42** [comment]
I can not reproduce the problem :(, it works for me.

**Comment by kelson42** [comment]
I suspect a kind of weakness in the code appearing only under certain conditions:
https://github.com/caolan/async/issues/75

**Comment by kelson42** [comment]
without a reproduce case, it's impossible to fix this bug.

**Comment by mdholloway** [comment]
I ran into this quite frequently at first.  I've worked around it locally simply by altering the behavior of 'speed' to represent the absolute concurrency rather than making it a multiple of the number of cores (with cpuCores remaining as the default), i.e.,

  var speed = argv.speed || cpuCount;

That allows me to reduce the concurrency to, say, half the number of cores, which seems to allow the script to complete reliably even for large (~50,000+ article) jobs.

Is that a change you'd consider merging?

**Comment by mdholloway** [comment]
This seems environment-specific and I don't know why.  I've only encountered it on a Wikimedia Cloud VPS instance (running Debian Jessie).  I can't reproduce it in any other context, including a personal laptop running Jessie.  (The v8 stack_size setting is the same in each case: 984 kb.)

In any case, I kept hitting this yesterday even notwithstanding my 'speed' hack, so I dug in a little more.  I noticed that this always occurs when processing anchor element attributes in the rewriteUrls loop; what seems to be happening is that the frequent iterative callbacks in that loop are overflowing the call stack as described in the "Synchronous iteration functions" section of the "Common Pitfalls" here:

https://caolan.github.io/async/

Wrapping the calls to finished() in setImmediate() callbacks, as suggested there, seems to relieve the pressure on the call stack and resolve the issue.  I'll do a little more testing and submit a PR if it continues to be effective.
